### PR TITLE
Fix #5397: Visit local method in Tailrec

### LINKED ISSUE #5397: @tailrec not properly enforced
```scala
class Test {
  def cond: Boolean = ???
  def foo(x: => Int) = ???

  @annotation.tailrec
  final def bar: Int =
    if (cond) foo(bar) // should be an error
    else bar
}
```

[t4649](https://github.com/lampepfl/dotty/blob/master/tests/pos/tailcall/t4649.scala) is a pending test case

### PR DESCRIPTION
**Pull request by allanrenucci**
If a method is annotated with @tailrec we now report invalid recursive
call within local method. This means we now visit by-name arguments,
closures and lifted `try`s

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/transform/TailRec.scala [hunk 1]
```diff
@@ -407,8 +408,12 @@ class TailRec extends MiniPhase {
           assert(false, "We should never have gotten inside a pattern")
           tree
 
-        case _: ValDef | _: DefDef | _: Super | _: This |
-             _: Literal | _: TypeTree | _: TypeDef | EmptyTree =>
+        case tree: DefDef =>
+          noTailTransform(tree.rhs)
+          tree
+
+        case _: ValDef | _: Super | _: This | _: Literal |
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/transform/TailRec.scala]: Why aren't ValDefs traversed ?

Comment (allanrenucci) [reply compiler/src/dotty/tools/dotc/transform/TailRec.scala]: This is not new code. My guess is that they can’t be in tail position. But we should probably traverse them for the same reasons we want to traverse local DefDef

File: tests/neg/i5397.scala [hunk 2]
```diff
@@ -0,0 +1,41 @@
+import annotation.tailrec
+
+object Test {
+  def foo(x: => Int) = 1
+  def bar(x: Int => Int) = 1
+
+  @tailrec def rec1: Int =
+    foo(rec1) // error not in tail position
```
Comment (sjrd) [inline tests/neg/i5397.scala]: Isn't the convention to start with `error:`, including the colon?

Comment (allanrenucci) [reply tests/neg/i5397.scala]: Indeed

**Pull request merged**